File: 2065/main.go
package main

import (
	"fmt"
)

func main() {
	fmt.Println(maximalPathQuality([]int{0, 32, 10, 43}, [][]int{
		{0, 1, 10},
		{1, 2, 15},
		{0, 3, 10},
	}, 49))
}

func maximalPathQuality(values []int, edges [][]int, maxTime int) int {
	n := len(values)
	g := make([][][2]int, n)
	for _, edge := range edges {
		g[edge[0]] = append(g[edge[0]], [2]int{edge[1], edge[2]})
		g[edge[1]] = append(g[edge[1]], [2]int{edge[0], edge[2]})
	}

	visited := make([]bool, n)
	visited[0] = true
	ans := 0

	var dfs func(int, int, int)
	dfs = func(u, time, value int) {
		if u == 0 {
			ans = max(ans, value)
		}
		for _, e := range g[u] {
			v, dist := e[0], e[1]
			if time + dist <= maxTime {
				if !visited[v] {
					visited[v] = true
					dfs(v, time + dist, value + values[v])
					visited[v] = false
				} else {
					dfs(v, time + dist, value)
				}
			}
		}
	}
	dfs(0, 0, values[0])
	return ans
}

File: 2065/main.rs
use std::cmp::max;

struct Solution{}

impl Solution {
    pub fn maximal_path_quality(values: Vec<i32>, edges: Vec<Vec<i32>>, max_time: i32) -> i32 {
        let n = values.len();
        let mut g:Vec<Vec<(i32, i32)>> = vec![vec![]; n];
        for edge in edges {
            g[edge[0] as usize].push((edge[1], edge[2]));
            g[edge[1] as usize].push((edge[0], edge[2]));
        }

        let mut visited = vec![false; n];
        visited[0] = true;
        let mut ans = 0;

        fn dfs(u: usize, time: i32, value: i32, ans: &mut i32, g: &Vec<Vec<(i32, i32)>>, visited: &mut Vec<bool>, values: &Vec<i32>, max_time: &i32) {
            if u == 0 {
                *ans = max(*ans, value);
            }
            for (v, dist) in &g[u] {

                if time + dist <= *max_time {
                    if !visited[*v as usize] {
                        visited[*v as usize] = true;
                        dfs(*v as usize, time+dist, value+values[*v as usize], ans, g, visited, values, max_time);
                        visited[*v as usize] = false;
                    } else {
                        dfs(*v as usize, time+dist, value, ans, g, visited, values, max_time);
                    }
                }
            }
        }
        dfs(0, 0, values[0], &mut ans, &g, &mut visited, &values, &max_time);
        ans
    }
}

fn main() {
	println!("{}", Solution::maximal_path_quality(vec![0, 32, 10, 43],vec![
		vec![0, 1, 10],
		vec![1, 2, 15],
		vec![0, 3, 10],
        ], 49))
}

